heartbeat/heartbeat_test.go

```diff
@@ -0,0 +1,204 @@
+	"os"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/choria-io/stream-replicator/config"
+	"github.com/choria-io/stream-replicator/election"
+	"github.com/choria-io/stream-replicator/internal/testutil"
+	"github.com/nats-io/jsm.go"
+	"github.com/nats-io/nats.go"
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+	"github.com/prometheus/client_golang/prometheus"
+	dto "github.com/prometheus/client_model/go"
+
+	"github.com/sirupsen/logrus"
+)
+
+func TestHeartBeat(t *testing.T) {
+	RegisterFailHandler(Fail)
+	RunSpecs(t, "heartbeat")
+}
+
+var _ = Describe("Subject Heartbeat", func() {
+	var (
+		ctx      context.Context
+		cancel   context.CancelFunc
+		wg       = sync.WaitGroup{}
+		log      *logrus.Entry
+		hbConfig config.HeartBeat
+	)
+
+	BeforeEach(func() {
+		ctx, cancel = context.WithTimeout(context.Background(), 30*time.Second)
+		logger := logrus.New()
+		logger.SetOutput(GinkgoWriter)
+		log = logrus.NewEntry(logger)
+		log.Logger.SetLevel(logrus.DebugLevel)
+
+		DeferCleanup(func() {
+			cancel()
+			wg.Wait()
+		})
+
+		hbConfig = config.HeartBeat{
+			LeaderElection: false,
+			Subjects: []config.Subject{
+				{
+					Name:     "heartbeat",
+					Interval: "500ms",
+				},
+			},
+		}
+		election.SkipTTLValidateForTests()
+	})
+
+	// for use with Eventually()
+	streamMesssage := func(s *jsm.Stream) func() (uint64, error) {
+		return func() (uint64, error) {
+			nfo, err := s.State()
+			if err != nil {
+				return 0, err
+			}
+			return nfo.Msgs, nil
+		}
+	}
+
+	Describe("run", func() {
+		It("should send at least 1 heart beat message", func() {
+			testutil.WithJetStream(log, func(nc *nats.Conn, mgr *jsm.Manager) {
+				jstream, err := mgr.NewStream("TEST", jsm.Subjects("heartbeat"))
+				Expect(err).ToNot(HaveOccurred())
+				hbConfig.URL = nc.ConnectedUrl()
+
+				hb, err := New(&hbConfig, "test_replicator", log)
+				Expect(err).ToNot(HaveOccurred())
+
+				go func() {
+					defer GinkgoRecover()
+					err = hb.Run(ctx, &wg)
+					Expect(err).ToNot(HaveOccurred())
+				}()
+				defer cancel()
+				Eventually(streamMesssage(jstream)).Should(BeNumerically(">=", 1))
+			})
+		})
+
+		It("should send a well formed message", func() {
+			testutil.WithJetStream(log, func(nc *nats.Conn, mgr *jsm.Manager) {
+				jstream, err := mgr.NewStream("TEST", jsm.Subjects("heartbeat"))
+				Expect(err).ToNot(HaveOccurred())
+				hbConfig.URL = nc.ConnectedUrl()
+				hostname, _ := os.Hostname()
+				hbConfig.Headers = map[string]string{
+					"foo": "bar",
+				}
+
+				hb, err := New(&hbConfig, "test_replicator", log)
+				Expect(err).ToNot(HaveOccurred())
+
+				go func() {
+					defer GinkgoRecover()
+					err = hb.Run(ctx, &wg)
+					Expect(err).ToNot(HaveOccurred())
+				}()
+				defer cancel()
+				Eventually(streamMesssage(jstream)).Should(BeNumerically(">=", 1))
+
+				x, err := nc.JetStream()
+				sub, err := x.PullSubscribe("heartbeat", "")
+				msgs, err := sub.Fetch(1)
+				Expect(msgs[0].Data).To(MatchRegexp("^\\d+$"))
```

Comment:
check it parses as timestamp and `BeNumerically("~", 10*time.Second, 10*time.Millisecond)`